pkg/gather/clusterconfig/clusterconfig.go

```diff
@@ -147,6 +148,99 @@ func GatherMostRecentMetrics(i *Gatherer) func() ([]record.Record, []error) {
 	}
 }
 
+// MinimalNodeInfo contains the most essential information about a node
+type MinimalNodeInfo struct {
+	ProviderID string `json:"providerID"`
+	Image      string `json:"image"`
+}
+
+// ContainerInstances map containing the number of running instances for a cluster name.
+type ContainerInstances map[string]int
+
+// ImageInstanceInfo hold information about pods running the given image.
+type ImageInstanceInfo struct {
+	Image     string
+	Instances ContainerInstances
+}
+
+// RunningImages assigns information about running containers to a specific image index.
+// The index is a reference to an item in the related `ContainerImageSet` instance.
+type RunningImages map[int]ContainerInstances
+
+// PodsWithAge maps the YYYY-MM string representation of start time to list of pods running since that month.
+type PodsWithAge map[string]RunningImages
+
+// Add inserts the specified container information into the data structure.
+func (p *PodsWithAge) Add(startTime time.Time, image int, containerName string) {
+	const YyyyMmFormat = "2006-01"
+	month := startTime.UTC().Format(YyyyMmFormat)
+	if imageMap, exists := (*p)[month]; exists {
+		if containerMap, exists := imageMap[image]; exists {
+			if _, exists := containerMap[containerName]; exists {
+				containerMap[containerName]++
+			} else {
+				containerMap[containerName] = 1
+			}
+		} else {
+			imageMap[image] = ContainerInstances{containerName: 1}
+		}
+	} else {
+		(*p)[month] = RunningImages{image: ContainerInstances{containerName: 1}}
+	}
+}
+
+// ContainerImageSet is used to store unique container image URLs.
+// The key is a continuous index starting from 0.
+// The value is the image URL itself.
+type ContainerImageSet map[int]string
+
+// Add ensures that the image set contains the specified index and returns its index.
+func (is *ContainerImageSet) Add(image string) int {
+	for idx, img := range *is {
+		if img == image {
+			return idx
+		}
+	}
+
+	nextIndex := len(*is)
+	(*is)[nextIndex] = image
+	return nextIndex
+}
+
+// ContainerInfo encapsulates the essential information about running containers in a minimalized data structure.
+type ContainerInfo struct {
+	Images     ContainerImageSet `json:"images"`
+	Containers PodsWithAge       `json:"containers"`
+}
+
+// GatherContainerImages collects essential information about running containers.
+// Specifically, the age of pods, the set of running images and the container names are collected.
+func GatherContainerImages(i *Gatherer) func() ([]record.Record, []error) {
+	return func() ([]record.Record, []error) {
+		pods, err := i.coreClient.Pods("").List(metav1.ListOptions{})
+		if err != nil {
+			return nil, []error{err}
+		}
+
+		contInfo := ContainerInfo{
+			Images:     ContainerImageSet{},
+			Containers: PodsWithAge{},
+		}
+
+		for _, pod := range pods.Items {
+			for _, container := range pod.Spec.Containers {
+				imgIndex := contInfo.Images.Add(container.Image)
```

Comment:
Ironically, this was the reason why I originally wanted the map to be inside-out - because it would make writing to it easier, but the lookup would be slower and probably a bit confusing as well.
I'll try to add something like that, but I'm worried there will be too many tricks like this and the code will become difficult to navigate at that point.